{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/therealhaseeb/MyThesis/blob/main/1_GenomicPipelineForThesis(Hail%2BPandas_Dask).ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Author: HASEEB ZAHID"
      ],
      "metadata": {
        "id": "n2X6sGaCv6Ec"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "uCQ2allqksQe",
        "outputId": "03816ffc-3343-4a98-c510-7a7365bfac70"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Requirement already satisfied: urllib3 in /usr/local/lib/python3.7/dist-packages (1.26.9)\n",
            "\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001b[0m\u001b[33m\n",
            "\u001b[0m"
          ]
        }
      ],
      "source": [
        "!pip install --upgrade urllib3"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "TpVJL_Hp2Jrt",
        "outputId": "013c2785-75b0-42ef-d762-7fe82da43d6d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "_St33MihUDfU"
      },
      "outputs": [],
      "source": [
        "!curl -O ftp://ngs.sanger.ac.uk/production/hgdp/hgdp_wgs.20190516/hgdp_wgs.20190516.full.chrX.vcf.gz\n",
        "!curl -O ftp://ngs.sanger.ac.uk/production/hgdp/hgdp_wgs.20190516/hgdp_wgs.20190516.full.chrX.vcf.gz.tbi"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "-ojuHIJ7Rktv"
      },
      "outputs": [],
      "source": [
        "!apt-get install -y \\\n",
        "    openjdk-8-jre-headless \\\n",
        "    g++ \\\n",
        "    python3.6 python3-pip \\\n",
        "    libopenblas-base liblapack3\n",
        "\n",
        "!pip install hail"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "he7WrHfFp88l",
        "outputId": "2caf0278-c998-43dd-9b2b-34673cd19d5b"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Initializing Hail with default parameters...\n",
            "Running on Apache Spark version 3.1.3\n",
            "SparkUI available at http://78fac6c4aa73:4040\n",
            "Welcome to\n",
            "     __  __     <>__\n",
            "    / /_/ /__  __/ /\n",
            "   / __  / _ `/ / /\n",
            "  /_/ /_/\\_,_/_/_/   version 0.2.97-937922d7f46c\n",
            "LOGGING: writing to /content/hail-20220725-1554-0.2.97-937922d7f46c.log\n",
            "2022-07-25 15:54:53 Hail: INFO: Coerced sorted dataset\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "(132457, 610)\n"
          ]
        }
      ],
      "source": [
        "mt = hl.import_vcf('/content/hgdp_wgs.20190516.full.chrX.vcf.gz', reference_genome='GRCh38', force_bgz=True)\n",
        "print(mt.count())"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "I2jKiBL8RqGA"
      },
      "outputs": [],
      "source": [
        "import hail as hl\n",
        "from hail.plot import output_notebook, show\n",
        "import bokeh\n",
        "from pprint import pprint\n",
        "import os\n",
        "import pandas as pd"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "IMyTsSbwuxLy",
        "outputId": "e49f63ad-c9e0-43ef-ba77-3a71532f4ce3"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading\n",
            "\n",
            "Applying Mutual Info Algorithm\n",
            "[0.06308505 0.1248506  0.00060279 ... 0.         0.         0.0913441 ]\n",
            "\n",
            "Saving outputs\n",
            "\n",
            "Done\n"
          ]
        }
      ],
      "source": [
        "%%capture cap\n",
        "\n",
        "chrom_list = ( list(range(1,22)) + ['X', 'Y'])\n",
        "\n",
        "table = (hl.import_table('/content/drive/MyDrive/thesis/Mutual Info Results/HGDP-Pak.txt', impute=True).key_by('Sample name'))\n",
        "import dask.dataframe as pd\n",
        "pheno = pd.read_csv(\"/content/drive/MyDrive/thesis/Mutual Info Results/HGDP-Pak.txt\", delimiter='\\t')\n",
        "\n",
        "# due to memory constraints, this pipeline is run on each chromosome one by one separately\n",
        "for chr in ['Y']:\n",
        "    try:\n",
        "        print(\"\\nReading Chromosome\", chr)\n",
        "        mydata = hl.import_vcf('/content/hgdp_wgs.20190516.full.chr'+str(chr)+'.vcf.gz', reference_genome='GRCh38', force_bgz=True)\n",
        "        mt = mydata.annotate_cols(pheno = table[mydata.s])\n",
        "\n",
        "        print('\\nCount;', mt.count())\n",
        "        mt2 = mt.filter_cols(mt.pheno['Population name'] != '')\n",
        "        print('\\nTaking Pakistani Population;', mt2.count())\n",
        "\n",
        "        mt2 = mt2.drop('PL')\n",
        "\n",
        "        print(\"\\nSplitting into Biallelics;\")\n",
        "        mt3 = hl.split_multi_hts(mt2)\n",
        "        print(\"\\nWriting\")\n",
        "        mt3.write(\"/content/drive/MyDrive/thesis/hgdp-\"+str(chr)+\"_split.mt\")\n",
        "        mt3 = hl.read_matrix_table(\"/content/drive/MyDrive/thesis/hgdp-\"+str(chr)+\"_split.mt\")\n",
        "\n",
        "        print(\"\\nLD Pruning;\")\n",
        "        pruned_variant_table = hl.ld_prune(mt3.GT, r2=0.2, bp_window_size=50000)\n",
        "        filtered_mt = mt3.filter_rows(hl.is_defined(pruned_variant_table[mt3.row_key]))\n",
        "        filtered_mt.write(\"/content/drive/MyDrive/thesis/hgdp-\"+str(chr)+\"_split_ldpruned.mt\", overwrite=True)\n",
        "        os.system(\"rm -r /content/drive/MyDrive/thesis/hgdp-\"+str(chr)+\"_split.mt\")\n",
        "\n",
        "        filtered_mt = hl.read_matrix_table(\"/content/drive/MyDrive/thesis/hgdp-\"+str(chr)+\"_split_ldpruned.mt\")\n",
        "        print(filtered_mt.count())\n",
        "\n",
        "        print(\"\\nExporting Genotypes\")\n",
        "        filtered_mt.GT.export(\"/content/drive/MyDrive/thesis/HGDP-chr\"+str(chr)+\"-GT-Samples_split&ldpruned.tsv\")\n",
        "\n",
        "        print(\"Deleting VCFs\")\n",
        "        os.system('rm hgdp_wgs.20190516.full.chr'+str(chr)+'.vcf.gz.tbi')\n",
        "        os.system('rm hgdp_wgs.20190516.full.chr'+str(chr)+'.vcf.gz')\n",
        "        os.system(\"rm -r /content/drive/MyDrive/thesis/hgdp-\"+str(chr)+\"_split.mt\")\n",
        "        os.system(\"/content/drive/MyDrive/thesis/hgdp-\"+str(chr)+\"_split_ldpruned.mt\")\n",
        "\n",
        "        print(\"Reading\")\n",
        "        df_gt1 = pd.read_csv(\"/content/drive/MyDrive/thesis/HGDP-chrX-GT-Samples_split&ldpruned.tsv\",\n",
        "                             delimiter='\\t',\n",
        "                             low_memory=False\n",
        "                             )\n",
        "\n",
        "        df_gt1['bp'] = df_gt1['locus'] + ':' + df_gt1['alleles']\n",
        "        del df_gt1['locus']\n",
        "        del df_gt1['alleles']\n",
        "\n",
        "        df_gt1 = df_gt1.set_index(\"bp\").T\n",
        "        print(\"\\nSaving GT csv with new column and transposing\")\n",
        "        df_gt1.fillna(3, inplace=True)\n",
        "        df_gt1.replace('0/0', 0, inplace=True)\n",
        "        df_gt1.replace('0/1', 1, inplace=True)\n",
        "        df_gt1.replace('1/0', 1, inplace=True)\n",
        "        df_gt1.replace('1/1', 2, inplace=True)\n",
        "        df_gt1.to_csv(\"/content/drive/MyDrive/thesis/HGDP Chromosomes Results/HGDP-chr\"+str(chr)+\"-GT-Samples_split&ldpruned&T.csv\")\n",
        "\n",
        "        print(\"\\nApplying Mutual Info Algorithm\")\n",
        "        df_gt1= pd.read_csv(\"/content/drive/MyDrive/thesis/HGDP Chromosomes Results/HGDP-chr\"+str(chr)+\"-GT-Samples_split&ldpruned&T.csv\")\n",
        "\n",
        "        merged_left = pd.merge(left=df_gt1, right=pheno, how='left', left_on='Unnamed: 0', right_on='Sample name')\n",
        "\n",
        "\n",
        "        x = merged_left.drop(['Sex', 'Unnamed: 0', 'Population name', 'Sample name', 'Superpopulation name',\t'Population elastic ID', 'Data collections'], axis=1)\n",
        "        y = merged_left['Population name']\n",
        "\n",
        "        from sklearn.feature_selection import mutual_info_classif as MIC\n",
        "        mi_score = MIC(x,y)\n",
        "        print(mi_score)\n",
        "\n",
        "        print(\"\\nSaving outputs\")\n",
        "        mi_scores_df = pd.DataFrame({\n",
        "            'Columns':x.columns,\n",
        "            'Scores':mi_score})\n",
        "        mi_scores_df.sort_values(ascending=False, by='Scores').to_csv(\"/content/drive/MyDrive/thesis/Mutual Info Results/hgdp-\"+str(chr)+\"_mi_scores.csv\")\n",
        "\n",
        "        print(\"\\nDone\")\n",
        "\n",
        "    except Exception as e: print(e)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "OoYBn3QEpRBh"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "j_FVGzcuZz3r",
        "outputId": "b5a386fa-2d85-41b2-e283-b4548673db0a"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-091acec6-a792-4467-bf24-592722548026\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Unnamed: 0</th>\n",
              "      <th>chrY:2781761:[\"CA\",\"C\"]</th>\n",
              "      <th>chrY:2781761:[\"CAA\",\"C\"]</th>\n",
              "      <th>chrY:2782347:[\"A\",\"G\"]</th>\n",
              "      <th>chrY:2782439:[\"G\",\"C\"]</th>\n",
              "      <th>chrY:2782986:[\"C\",\"T\"]</th>\n",
              "      <th>chrY:2785100:[\"T\",\"C\"]</th>\n",
              "      <th>chrY:2786162:[\"C\",\"G\"]</th>\n",
              "      <th>chrY:2788215:[\"T\",\"C\"]</th>\n",
              "      <th>chrY:2789135:[\"C\",\"T\"]</th>\n",
              "      <th>...</th>\n",
              "      <th>chrY:56887316:[\"G\",\"A\"]</th>\n",
              "      <th>chrY:56887463:[\"C\",\"A\"]</th>\n",
              "      <th>chrY:56887503:[\"G\",\"T\"]</th>\n",
              "      <th>chrY:56887583:[\"G\",\"A\"]</th>\n",
              "      <th>Sample name</th>\n",
              "      <th>Sex</th>\n",
              "      <th>Population name</th>\n",
              "      <th>Superpopulation name</th>\n",
              "      <th>Population elastic ID</th>\n",
              "      <th>Data collections</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>HGDP00001</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>HGDP00001</td>\n",
              "      <td>male</td>\n",
              "      <td>Brahui</td>\n",
              "      <td>Central South Asia (HGDP)</td>\n",
              "      <td>BrahuiHGDP</td>\n",
              "      <td>Human Genome Diversity Project</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>HGDP00003</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>HGDP00003</td>\n",
              "      <td>male</td>\n",
              "      <td>Brahui</td>\n",
              "      <td>Central South Asia (HGDP)</td>\n",
              "      <td>BrahuiHGDP</td>\n",
              "      <td>Human Genome Diversity Project</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>HGDP00005</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>HGDP00005</td>\n",
              "      <td>male</td>\n",
              "      <td>Brahui</td>\n",
              "      <td>Central South Asia (HGDP)</td>\n",
              "      <td>BrahuiHGDP</td>\n",
              "      <td>Human Genome Diversity Project</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>HGDP00007</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>HGDP00007</td>\n",
              "      <td>male</td>\n",
              "      <td>Brahui</td>\n",
              "      <td>Central South Asia (HGDP)</td>\n",
              "      <td>BrahuiHGDP</td>\n",
              "      <td>Human Genome Diversity Project</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>HGDP00009</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>HGDP00009</td>\n",
              "      <td>male</td>\n",
              "      <td>Brahui</td>\n",
              "      <td>Central South Asia (HGDP)</td>\n",
              "      <td>BrahuiHGDP</td>\n",
              "      <td>Human Genome Diversity Project</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>148</th>\n",
              "      <td>HGDP00417</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>HGDP00417</td>\n",
              "      <td>male</td>\n",
              "      <td>Burusho</td>\n",
              "      <td>Central South Asia (HGDP)</td>\n",
              "      <td>BurushoHGDP</td>\n",
              "      <td>Human Genome Diversity Project</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>149</th>\n",
              "      <td>HGDP00423</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>HGDP00423</td>\n",
              "      <td>male</td>\n",
              "      <td>Burusho</td>\n",
              "      <td>Central South Asia (HGDP)</td>\n",
              "      <td>BurushoHGDP</td>\n",
              "      <td>Human Genome Diversity Project</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>150</th>\n",
              "      <td>HGDP00433</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>HGDP00433</td>\n",
              "      <td>male</td>\n",
              "      <td>Burusho</td>\n",
              "      <td>Central South Asia (HGDP)</td>\n",
              "      <td>BurushoHGDP</td>\n",
              "      <td>Human Genome Diversity Project</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>151</th>\n",
              "      <td>HGDP00438</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>HGDP00438</td>\n",
              "      <td>male</td>\n",
              "      <td>Burusho</td>\n",
              "      <td>Central South Asia (HGDP)</td>\n",
              "      <td>BurushoHGDP</td>\n",
              "      <td>Human Genome Diversity Project</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>152</th>\n",
              "      <td>HGDP00445</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>1.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1.0</td>\n",
              "      <td>HGDP00445</td>\n",
              "      <td>male</td>\n",
              "      <td>Burusho</td>\n",
              "      <td>Central South Asia (HGDP)</td>\n",
              "      <td>BurushoHGDP</td>\n",
              "      <td>Human Genome Diversity Project</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>153 rows × 20509 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-091acec6-a792-4467-bf24-592722548026')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-091acec6-a792-4467-bf24-592722548026 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-091acec6-a792-4467-bf24-592722548026');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "    Unnamed: 0  chrY:2781761:[\"CA\",\"C\"]  chrY:2781761:[\"CAA\",\"C\"]  \\\n",
              "0    HGDP00001                      0.0                       0.0   \n",
              "1    HGDP00003                      0.0                       0.0   \n",
              "2    HGDP00005                      0.0                       0.0   \n",
              "3    HGDP00007                      0.0                       1.0   \n",
              "4    HGDP00009                      NaN                       NaN   \n",
              "..         ...                      ...                       ...   \n",
              "148  HGDP00417                      0.0                       0.0   \n",
              "149  HGDP00423                      1.0                       0.0   \n",
              "150  HGDP00433                      1.0                       0.0   \n",
              "151  HGDP00438                      1.0                       0.0   \n",
              "152  HGDP00445                      0.0                       0.0   \n",
              "\n",
              "     chrY:2782347:[\"A\",\"G\"]  chrY:2782439:[\"G\",\"C\"]  chrY:2782986:[\"C\",\"T\"]  \\\n",
              "0                       0.0                     0.0                     0.0   \n",
              "1                       0.0                     0.0                     0.0   \n",
              "2                       0.0                     0.0                     0.0   \n",
              "3                       0.0                     0.0                     0.0   \n",
              "4                       0.0                     0.0                     0.0   \n",
              "..                      ...                     ...                     ...   \n",
              "148                     0.0                     0.0                     0.0   \n",
              "149                     0.0                     0.0                     0.0   \n",
              "150                     0.0                     0.0                     0.0   \n",
              "151                     0.0                     0.0                     0.0   \n",
              "152                     0.0                     0.0                     0.0   \n",
              "\n",
              "     chrY:2785100:[\"T\",\"C\"]  chrY:2786162:[\"C\",\"G\"]  chrY:2788215:[\"T\",\"C\"]  \\\n",
              "0                       0.0                     0.0                     0.0   \n",
              "1                       0.0                     0.0                     0.0   \n",
              "2                       0.0                     0.0                     0.0   \n",
              "3                       0.0                     0.0                     0.0   \n",
              "4                       0.0                     0.0                     0.0   \n",
              "..                      ...                     ...                     ...   \n",
              "148                     0.0                     0.0                     0.0   \n",
              "149                     1.0                     0.0                     0.0   \n",
              "150                     0.0                     0.0                     0.0   \n",
              "151                     0.0                     0.0                     0.0   \n",
              "152                     0.0                     0.0                     0.0   \n",
              "\n",
              "     chrY:2789135:[\"C\",\"T\"]  ...  chrY:56887316:[\"G\",\"A\"]  \\\n",
              "0                       1.0  ...                      0.0   \n",
              "1                       0.0  ...                      0.0   \n",
              "2                       0.0  ...                      0.0   \n",
              "3                       0.0  ...                      0.0   \n",
              "4                       0.0  ...                      0.0   \n",
              "..                      ...  ...                      ...   \n",
              "148                     0.0  ...                      0.0   \n",
              "149                     0.0  ...                      0.0   \n",
              "150                     0.0  ...                      0.0   \n",
              "151                     0.0  ...                      0.0   \n",
              "152                     0.0  ...                      1.0   \n",
              "\n",
              "     chrY:56887463:[\"C\",\"A\"]  chrY:56887503:[\"G\",\"T\"]  \\\n",
              "0                        0.0                      0.0   \n",
              "1                        0.0                      0.0   \n",
              "2                        0.0                      0.0   \n",
              "3                        1.0                      0.0   \n",
              "4                        0.0                      0.0   \n",
              "..                       ...                      ...   \n",
              "148                      0.0                      0.0   \n",
              "149                      0.0                      0.0   \n",
              "150                      0.0                      0.0   \n",
              "151                      1.0                      0.0   \n",
              "152                      NaN                      NaN   \n",
              "\n",
              "     chrY:56887583:[\"G\",\"A\"]  Sample name   Sex  Population name  \\\n",
              "0                        0.0    HGDP00001  male           Brahui   \n",
              "1                        1.0    HGDP00003  male           Brahui   \n",
              "2                        0.0    HGDP00005  male           Brahui   \n",
              "3                        1.0    HGDP00007  male           Brahui   \n",
              "4                        1.0    HGDP00009  male           Brahui   \n",
              "..                       ...          ...   ...              ...   \n",
              "148                      1.0    HGDP00417  male          Burusho   \n",
              "149                      1.0    HGDP00423  male          Burusho   \n",
              "150                      0.0    HGDP00433  male          Burusho   \n",
              "151                      1.0    HGDP00438  male          Burusho   \n",
              "152                      1.0    HGDP00445  male          Burusho   \n",
              "\n",
              "          Superpopulation name  Population elastic ID  \\\n",
              "0    Central South Asia (HGDP)             BrahuiHGDP   \n",
              "1    Central South Asia (HGDP)             BrahuiHGDP   \n",
              "2    Central South Asia (HGDP)             BrahuiHGDP   \n",
              "3    Central South Asia (HGDP)             BrahuiHGDP   \n",
              "4    Central South Asia (HGDP)             BrahuiHGDP   \n",
              "..                         ...                    ...   \n",
              "148  Central South Asia (HGDP)            BurushoHGDP   \n",
              "149  Central South Asia (HGDP)            BurushoHGDP   \n",
              "150  Central South Asia (HGDP)            BurushoHGDP   \n",
              "151  Central South Asia (HGDP)            BurushoHGDP   \n",
              "152  Central South Asia (HGDP)            BurushoHGDP   \n",
              "\n",
              "                   Data collections  \n",
              "0    Human Genome Diversity Project  \n",
              "1    Human Genome Diversity Project  \n",
              "2    Human Genome Diversity Project  \n",
              "3    Human Genome Diversity Project  \n",
              "4    Human Genome Diversity Project  \n",
              "..                              ...  \n",
              "148  Human Genome Diversity Project  \n",
              "149  Human Genome Diversity Project  \n",
              "150  Human Genome Diversity Project  \n",
              "151  Human Genome Diversity Project  \n",
              "152  Human Genome Diversity Project  \n",
              "\n",
              "[153 rows x 20509 columns]"
            ]
          },
          "execution_count": 11,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "merged_left"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Eb9ruVyvklhi"
      },
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "mount_file_id": "14eRtJBLWavQul5YjGrU3_B8Ihkhe0cfG",
      "authorship_tag": "ABX9TyPY9379Ynu7eG/7HnXKaJkq",
      "include_colab_link": true
    },
    "gpuClass": "standard",
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}